File: server/handlers/sessions.py
"""
Session and projects HTTP handlers.

Delegates to SessionService; list_projects uses config + server.projects.
"""

from __future__ import annotations

from fastapi import HTTPException, Request

from server.models import (
    BranchSessionRequest,
    EditMessageRequest,
    RenameSessionRequest,
    SessionDetail,
    SessionListItem,
    SessionProjectPatch,
)


async def list_sessions(
    request: Request,
    auth_user: object,
    q: str | None = None,
) -> list[SessionListItem]:
    """List all chat sessions with optional search filter."""
    svc = request.app.state.session_service
    return svc.list_sessions(q=q)


async def get_session(
    session_id: str,
    request: Request,
    auth_user: object,
) -> SessionDetail:
    """Get full session history."""
    svc = request.app.state.session_service
    return svc.get_session(session_id)


async def delete_session(
    session_id: str,
    request: Request,
    auth_user: object,
) -> dict:
    """Permanently delete a session."""
    svc = request.app.state.session_service
    svc.delete_session(session_id)
    return {"status": "deleted", "key": session_id}


async def rename_session(
    session_id: str,
    body: RenameSessionRequest,
    request: Request,
    auth_user: object,
) -> dict:
    """Rename a session by setting title in metadata."""
    svc = request.app.state.session_service
    svc.rename_session(session_id, body.title)
    return {"status": "renamed", "key": session_id, "title": body.title}


async def new_session(
    session_id: str,
    request: Request,
    auth_user: object,
) -> dict:
    """Start a new conversation (clear history)."""
    svc = request.app.state.session_service
    svc.new_session(session_id)
    return {"status": "new_session", "key": session_id}


async def set_session_project(
    session_id: str,
    body: SessionProjectPatch,
    request: Request,
    auth_user: object,
) -> dict:
    """Set or clear the active project for this session."""
    svc = request.app.state.session_service
    project = svc.set_project(session_id, body.project)
    return {"status": "ok", "key": session_id, "project": project}


async def retry_session(
    session_id: str,
    request: Request,
    auth_user: object,
) -> dict:
    """Branch from last user message and return new key / message count."""
    svc = request.app.state.session_service
    status, key, message_count = svc.retry_session(session_id)
    return {"status": status, "key": key, "message_count": message_count}


async def branch_session(
    session_id: str,
    body: BranchSessionRequest,
    request: Request,
    auth_user: object,
) -> dict:
    """Branch from a specific message into a new session."""
    svc = request.app.state.session_service
    try:
        status, new_key, message_count = svc.branch_session(session_id, body.message_index)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"status": status, "key": new_key, "message_count": message_count}


async def edit_session_message(
    session_id: str,
    body: EditMessageRequest,
    request: Request,
    auth_user: object,
) -> dict:
    """Truncate session to message_index (exclusive); frontend sends new_content as new message."""
    svc = request.app.state.session_service
    status, key, message_count = svc.edit_session_message(
        session_id, body.message_index, body.new_content
    )
    return {"status": status, "key": key, "message_count": message_count}


async def list_projects(request: Request) -> list[dict]:
    """List projects from ~/.nanobot/projects.json for Project Context Switcher."""
    from server.projects import load_projects

    config = request.app.state.config
    projects = load_projects(config.workspace_path)
    return [{"name": k, "path": v} for k, v in sorted(projects.items())]

File: frontend/src/api/sessions.ts
/** Session API: list, get, delete, rename, new, branch, retry, edit, set project */

import { apiFetch } from './http'
import type { SessionDetail, SessionListItem } from './types/sessions'

/** List all sessions, sorted newest-first. Optional search query filters by title, chat_id, or key. */
export function getSessions(q?: string): Promise<SessionListItem[]> {
  const params = q?.trim() ? `?q=${encodeURIComponent(q.trim())}` : ''
  return apiFetch(`/sessions${params}`)
}

/** Get full history for a session. The key may contain ':', URL-encode it. */
export function getSession(key: string): Promise<SessionDetail> {
  return apiFetch(`/sessions/${encodeURIComponent(key)}`)
}

/** Permanently delete a session (removes file from disk). */
export function deleteSession(key: string): Promise<{ status: string; key: string }> {
  return apiFetch(`/sessions/${encodeURIComponent(key)}`, { method: 'DELETE' })
}

/** Rename a session by setting a human-readable title. */
export function renameSession(
  key: string,
  title: string,
): Promise<{ status: string; key: string; title: string }> {
  return apiFetch(`/sessions/${encodeURIComponent(key)}/rename`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ title }),
  })
}

/** Pop the last assistant message so the frontend can retry. */
export function retrySession(
  key: string,
): Promise<{ status: string; key: string; message_count: number }> {
  return apiFetch(`/sessions/${encodeURIComponent(key)}/retry`, { method: 'POST' })
}

/** Truncate the session to messageIndex (exclusive) so the frontend can re-send an edited user message. */
export function editSessionMessage(
  key: string,
  messageIndex: number,
  newContent: string,
): Promise<{ status: string; key: string; message_count: number }> {
  return apiFetch(`/sessions/${encodeURIComponent(key)}/edit`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ message_index: messageIndex, new_content: newContent }),
  })
}

/** Start a new conversation (clears history). */
export function newSession(key: string): Promise<{ status: string; key: string }> {
  return apiFetch(`/sessions/${encodeURIComponent(key)}/new`, { method: 'POST' })
}

/** Branch from a message: copy messages[0..messageIndex] into a new session. Returns the new session key. */
export function branchSession(
  key: string,
  messageIndex: number,
): Promise<{ status: string; key: string; message_count: number }> {
  return apiFetch(`/sessions/${encodeURIComponent(key)}/branch`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ message_index: messageIndex }),
  })
}

/** Set or clear the active project for a session. */
export function setSessionProject(
  sessionKey: string,
  project: string | null,
): Promise<{ status: string; key: string; project: string | null }> {
  return apiFetch(`/sessions/${encodeURIComponent(sessionKey)}/project`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ project }),
  })
}
